pandas/core/algorithms.py

```diff
@@ -706,6 +706,19 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
                 keys = Index(keys)
             result = Series(counts, index=keys, name=name)
 
+            # Use same index as the original values
+            if result.index.isna().sum() > 0:
```

Comment:
I don't understand this section. Are there cases where `values.fillna(np.nan)` does anything?

Reply:
What basically happens a few lines further is a reindex of a value_counts:
```
vals = [1, None, 2, 2]
vals.value_counts(dropna=False).reindex(pd.Series(pd.unique([1, None, 2, 2])))
```
The value_counts transforms the None into NaN in the index, so that the count will be NaN as well (and this value count is removed from the returning reindexed Series).

The fillna is used to transform the None from the reindex as well, so that this is executed instead, so that the resulting Series still contains the value count:
```
pd.Series(vals).value_counts(dropna=False).reindex(pd.Series(pd.unique(vals)).fillna(np.NaN)
```

I hope this makes sense, because I am quite confused as well from times. If you have another solution to fix this edge case, please let me know...

Reply:
@tomspur I am not sure what you are doing here at all. see the OP for the solution, which is a simple re-index. this PR is way out of scope.

Reply:
@jreback 
The re-index seems to work for `sort=False` and non-categorical indices and I could add that in this PR.

I wanted to further fix this for `sort=True`, when several values have the same count, e.g. this from the tests that were added in this PR:
```
In [2]: Series(list('bacaef'))
In [3]: s.value_counts(sort=True)                                                                                                                                                                                                             
Out[3]: 
a    2
c    1
e    1
b    1
f    1
dtype: int64
```
And I wanted to have them in the order `abcef` as well in that case.

I'll have a look then at `sort=False` for now as you suggested and leave `sort=True` for another PR